File: IOSAccessAssessment/SequentialProcessing/centroid_tracker.py
from collections import OrderedDict
import numpy as np


class CentroidTracker:
    def __init__(self, maxDisappeared=5):
        # initialize the next unique object ID along with two ordered
        # dictionaries used to keep track of mapping a given object
        # ID to its centroid and number of consecutive frames it has
        # been marked as "disappeared", respectively
        self.nextObjectID = 0
        self.objects = OrderedDict()
        self.object_class_and_distance = OrderedDict()
        self.disappeared = OrderedDict()

        # store the number of maximum consecutive frames a given
        # object is allowed to be marked as "disappeared" until we
        # need to deregister the object from tracking
        self.maxDisappeared = maxDisappeared

    def register(
        self,
        object_name,
        object_centroid,
        object_poly,
        object_distance,
        object_width,
        object_heading,
    ):
        # when registering an object we use the next available object
        # ID to store the centroid

        # self.objects[self.nextObjectID] = centroid
        object_dict = {}
        object_dict["object_name"] = object_name
        object_dict["object_centroid"] = object_centroid
        object_dict["object_poly"] = object_poly
        object_dict["object_distance"] = object_distance
        object_dict["object_width"] = object_width
        object_dict["object_heading"] = object_heading

        self.objects[self.nextObjectID] = object_dict
        self.disappeared[self.nextObjectID] = 0
        self.nextObjectID += 1

    def deregister(self, objectID):
        # to deregister an object ID we delete the object ID from
        # both of our respective dictionaries
        del self.objects[objectID]
        del self.disappeared[objectID]

    def update(self, object_list_dict):
        inputCentroids = object_list_dict["object_centroids"]
        inputPolys = object_list_dict["object_polys"]
        inputNames = object_list_dict["object_names"]
        inputDistances = object_list_dict["object_distances"]
        inputWidths = object_list_dict["object_widths"]
        inputHeadings = object_list_dict["headings"]

        # check to see if the list of input bounding box rectangles
        # is empty
        if len(inputCentroids) == 0:
            # loop over any existing tracked objects and mark them
            # as disappeared
            for objectID in list(self.disappeared.keys()):
                self.disappeared[objectID] += 1

                # if we have reached a maximum number of consecutive
                # frames where a given object has been marked as
                # missing, deregister it
                if self.disappeared[objectID] > self.maxDisappeared:
                    self.deregister(objectID)

            # return early as there are no centroids or tracking info
            # to update
            return self.objects, self.disappeared

        # if we are currently not tracking any objects take the input
        # centroids and register each of them
        if len(self.objects) == 0:
            for i in range(0, len(inputCentroids)):
                self.register(
                    inputNames[i],
                    inputCentroids[i],
                    inputPolys[i],
                    inputDistances[i],
                    inputWidths[i],
                    inputHeadings[i],
                )

        # otherwise, are are currently tracking objects so we need to
        # try to match the input centroids to existing object
        # centroids
        else:
            # grab the set of object IDs and corresponding centroids
            objectIDs = list(self.objects.keys())
            objectCentroids = list(
                object_dict["object_centroid"] for object_dict in self.objects.values()
            )

            # compute the distance between each pair of object
            # centroids and input centroids, respectively -- our
            # goal will be to match an input centroid to an existing
            # object centroid

            # Convert centroids to NumPy arrays
            objectCentroids = np.array(objectCentroids)
            inputCentroids = np.array(inputCentroids)

            # Compute the Euclidean distance between each pair of centroids
            D = np.linalg.norm(
                objectCentroids[:, np.newaxis, :] - inputCentroids[np.newaxis, :, :],
                axis=2,
            )

            # in order to perform this matching we must (1) find the
            # smallest value in each row and then (2) sort the row
            # indexes based on their minimum values so that the row
            # with the smallest value as at the *front* of the index
            # list
            rows = D.min(axis=1).argsort()

            # next, we perform a similar process on the columns by
            # finding the smallest value in each column and then
            # sorting using the previously computed row index list
            cols = D.argmin(axis=1)[rows]

            # in order to determine if we need to update, register,
            # or deregister an object we need to keep track of which
            # of the rows and column indexes we have already examined
            usedRows = set()
            usedCols = set()

            # loop over the combination of the (row, column) index
            # tuples
            for row, col in zip(rows, cols):
                # (existing object, input object)

                # threshold D to prevent merging far objects
                if D[row, col] > 50:
                    continue

                # if we have already examined either the row or
                # column value before, ignore it
                if row in usedRows or col in usedCols:
                    continue

                # otherwise, grab the object ID for the current row,
                # set its new centroid, and reset the disappeared
                # counter
                objectID = objectIDs[row]
                self.objects[objectID]["object_centroid"] = inputCentroids[col]
                self.objects[objectID]["object_name"] = inputNames[col]
                self.objects[objectID]["object_poly"] = inputPolys[col]
                self.objects[objectID]["object_distance"] = inputDistances[col]
                self.objects[objectID]["object_width"] = inputWidths[col]
                self.objects[objectID]["object_heading"] = inputHeadings[col]

                self.disappeared[objectID] = 0

                # indicate that we have examined each of the row and
                # column indexes, respectively
                usedRows.add(row)
                usedCols.add(col)

            # compute both the row and column index we have NOT yet
            # examined
            unusedRows = set(range(0, D.shape[0])).difference(usedRows)
            unusedCols = set(range(0, D.shape[1])).difference(usedCols)

            # in the event that the number of object centroids is
            # equal or greater than the number of input centroids
            # we need to check and see if some of these objects have
            # potentially disappeared
            if D.shape[0] >= D.shape[1]:
                # loop over the unused row indexes
                for row in unusedRows:
                    # grab the object ID for the corresponding row
                    # index and increment the disappeared counter
                    objectID = objectIDs[row]
                    self.disappeared[objectID] += 1

                    # check to see if the number of consecutive
                    # frames the object has been marked "disappeared"
                    # for warrants deregistering the object
                    if self.disappeared[objectID] > self.maxDisappeared:
                        self.deregister(objectID)

            # otherwise, if the number of input centroids is greater
            # than the number of existing object centroids we need to
            # register each new input centroid as a trackable object
            else:
                for col in unusedCols:
                    self.register(
                        inputNames[col],
                        inputCentroids[col],
                        inputPolys[col],
                        inputDistances[col],
                        inputWidths[col],
                        inputHeadings[col],
                    )

        # return the set of trackable objects
        return self.objects, self.disappeared

File: IOSAccessAssessment/Segmentation/PostProcessing/CentroidTracker.swift
//
//  CentroidTracker.swift
//  IOSAccessAssessment
//
//  Created by Himanshu on 4/7/25.
//

import OrderedCollections

struct TrackedObject {
    var classLabel: UInt8
    var centroid: CGPoint
    var polygon: [SIMD2<Float>] // Using SIMD2<Float> for polygon points because contour detection typically returns points in this format
    var boundingBox: CGRect?
//    var location: (Float, Float)
//    var heading: Float
//    var width: Float
}

class CentroidTracker {
    private var nextObjectID: UUID;
    private var objects: OrderedDictionary<UUID, TrackedObject>;
    private var disappearedObjects: OrderedDictionary<UUID, Int>;
    
    var maxDisappeared: Int;
    var distanceThreshold: Float;
    
    init(maxDisappeared: Int, distanceThreshold: Float = 50.0) {
        self.nextObjectID = UUID()
        self.objects = OrderedDictionary()
        self.disappearedObjects = OrderedDictionary()
        
        self.maxDisappeared = maxDisappeared
        self.distanceThreshold = distanceThreshold
    }
    
    func register(objectClassLabel: UInt8, objectCentroid: CGPoint, objectPolygon: Array<SIMD2<Float>>, objectBoundingBox: CGRect? = nil) {
        let object = TrackedObject(classLabel: objectClassLabel, centroid: objectCentroid, polygon: objectPolygon, boundingBox: objectBoundingBox)
        self.objects[nextObjectID] = object
        self.disappearedObjects[nextObjectID] = 0
        
        nextObjectID = UUID()
    }
    
    func deregister(objectID: UUID) {
        self.objects.removeValue(forKey: objectID)
        self.disappearedObjects.removeValue(forKey: objectID)
    }
    
    func update(objectsList: Array<TrackedObject>) ->
    (objects: OrderedDictionary<UUID, TrackedObject>, disappearedObjects: OrderedDictionary<UUID, Int>) {
        /**
         If object list is empty, increment the disappeared count for each object
         */
        if (objectsList.isEmpty) {
            for (objectID, objectDisappearCount) in self.disappearedObjects {
                self.disappearedObjects[objectID] = objectDisappearCount + 1;
                
                if (objectDisappearCount >= self.maxDisappeared) {
                    self.deregister(objectID: objectID);
                }
            }
            return (objects: self.objects, disappearedObjects: self.disappearedObjects)
        }
        
        /**
         If the current object list is empty, only register the new objects
         */
        if (self.objects.isEmpty) {
            for object in objectsList {
                self.register(objectClassLabel: object.classLabel, objectCentroid: object.centroid,
                              objectPolygon: object.polygon, objectBoundingBox: object.boundingBox);
            }
            return (objects: self.objects, disappearedObjects: self.disappearedObjects)
        }
        
        /**
         Otherwise, we need to match the objects in the current list with the existing objects
         */
        let objectIDs = Array(self.objects.keys);
        let objectCentroids = Array(self.objects.values.map { $0.centroid });
        
        let inputObjectCentroids = Array(objectsList.map { $0.centroid });
        
        /**
         Compute the distance matrix between the existing objects and the new objects.         
         */
        let distanceMatrix = computeDistanceMatrix(objectCentroids: objectCentroids, inputCentroids: inputObjectCentroids);
        let rowCount = distanceMatrix.count;
        let colCount = distanceMatrix[0].count;
        
        /**
         Get the row and column indices of the minimum distance in the distance matrix.
         */
        // Get the minimum distance in each row and store (rowIndex, minValue)
        let rowMinPairs: [(row: Int, minVal: Float)] = distanceMatrix.enumerated().map { (i, row) in
            return (i, row.min() ?? Float.infinity)
        }
        // Sort rows based on their min distance values (ascending)
        let rows: [Int] = rowMinPairs.sorted(by: { $0.minVal < $1.minVal }).map { $0.row }
        // For each sorted row, find the column index of the smallest distance
        let cols: [Int] = rows.map { rowIndex in
            distanceMatrix[rowIndex].enumerated().min(by: { $0.element < $1.element })?.offset ?? -1
        }
        let minPairs = zip(rows, cols)
        
        var usedRows = Set<Int>()
        var usedCols = Set<Int>()
        
        /**
            Loop through the minimum pairs and match the objects
         */
        for (row, col) in minPairs {
            // Check if the row and column are already used
            if (usedRows.contains(row) || usedCols.contains(col)) {
                continue
            }
            // Check if the distance is within the threshold
            if (distanceMatrix[row][col] > self.distanceThreshold) {
                continue
            }
            
            // Else, we have a match
            let objectID = objectIDs[row]
            self.objects[objectID] = objectsList[col]
            self.disappearedObjects[objectID] = 0
            
            usedRows.insert(row)
            usedCols.insert(col)
        }
        /**
         Get the unused rows and columns
         */
        let unusedRows = Set(0..<rowCount).subtracting(usedRows)
        let unusedCols = Set(0..<colCount).subtracting(usedCols)
        
        /**
         If there are any unused rows, increment the disappeared count for those objects and check if they need to be deregistered
         */
        for row in unusedRows {
            let objectID = objectIDs[row]
            self.disappearedObjects[objectID] = (self.disappearedObjects[objectID] ?? 0) + 1
            
            if ((self.disappearedObjects[objectID] ?? 0) >= self.maxDisappeared) {
                self.deregister(objectID: objectID)
            }
        }
        /**
         If there are any unused columns, register them as new objects
         */
        for col in unusedCols {
            let object = objectsList[col]
            self.register(objectClassLabel: object.classLabel, objectCentroid: object.centroid,
                          objectPolygon: object.polygon, objectBoundingBox: object.boundingBox)
        }
        
        return (objects: self.objects, disappearedObjects: self.disappearedObjects)
    }
    
    private func getCentroidDistance(centroid1: CGPoint, centroid2: CGPoint) -> Float {
        let dx = Float(centroid1.x - centroid2.x)
        let dy = Float(centroid1.y - centroid2.y)
        return sqrt(dx * dx + dy * dy)
    }
    
    private func computeDistanceMatrix(objectCentroids: [CGPoint], inputCentroids: [CGPoint]) -> [[Float]] {
        return objectCentroids.map { obj in
            inputCentroids.map { input in
                getCentroidDistance(centroid1: obj, centroid2: input)
            }
        }
    }
    
}
